subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

```diff
@@ -0,0 +1,94 @@
++
+image::deprecations.png[Deprecations View of a Gradle Build Scan]
++
+This is so that you can see any deprecation warnings that apply to your build.
++
+Alternatively, you could run `gradle help --warning-mode=all` to see the deprecations in the console, though it may not report as much detailed information.
+. Update your plugins.
++
+Some plugins will break with this new version of Gradle, for example because they use internal APIs that have been removed or changed. The previous step will help you identify potential problems by issuing deprecation warnings when a plugin does try to use a deprecated part of the API.
++
+. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
+. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.
+
+[[changes_9.0]]
+== Upgrading from 8.x and earlier
+
+=== Warnings that are now errors
+
+[[deprecated_configurations_should_not_be_used]]
+==== Usage of deprecated configurations is now an error
+
+Usage of the following `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` are now errors:
+
+- `LEGACY`
+- `DEPRECATED_CONSUMABLE`
+- `DEPRECATED_RESOLVABLE`
+
+=== Removed APIs
+
+=== Potential breaking changes
+
+=== Deprecations
+
+[[custom_configuration_roles]]
+Custom roles have been deprecated. Use pre-defined roles from `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` instead.
+
+[[reserved_configuration_names]]
+The `java-base` plugin will create configurations with for every sourceset using these patterns:
+
+ - `<SourceSet Name>Implementation`
+ - `<SourceSet Name>CompileOnly`
+ - `<SourceSet Name>RuntimeOnly`
+ - `<SourceSet Name>CompileClasspath`
+ - `<SourceSet Name>RuntimeClasspath`
+ - `<SourceSet Name>AnnotationProcessor`
+ - `<SourceSet Name>RuntimeElements`
+
+In addition, the `java-library` plugin will create configurations with these patterns:
+
+ - `<SourceSet Name>Api`
+ - `<SourceSet Name>ApiElements`
+ - `<SourceSet Name>CompileOnlyApi`
+
+The names of these standard configurations should be considered reserved - do not explicitly create configurations using them.
+
+See the <<java_plugin.adoc#tab:configurations, The Java Plugin's Dependency Management Reference>> for more information.
+
+[[upgrading_version_8:changes_8.0]]
+
+[[configurations_allowed_usage]]
+The usage of configurations should be fixed at creation.  Mutating the allowed usage on a configuration is deprecated.
+This includes calling any of the following `Configuration` methods:
+
+- `setCanBeConsumed(boolean)`
+- `setCanBeResolved(boolean)`
+
+It also includes calling the following method on `ConfigurationInternal`:
```

Comment:
Should we be calling out gradle internals in the upgrade guide?

Reply:
I'll remove it.